core/shared/src/main/scala/fs2/internal/Algebra.scala

```diff
@@ -99,9 +92,33 @@ private[fs2] object Algebra {
   def release[F[_], O](token: Token): FreeC[Algebra[F, O, ?], Unit] =
     FreeC.Eval[Algebra[F, O, ?], Unit](Release(token))
 
+  /**
+    * Steps through the stream, providing either `uncons` or `stepLeg`.
+    * Yields to head in form of segment, then id of the scope that was active after step evaluated and tail of the `stream`.
+    *
+    * @param stream             Stream to step
+    * @param scopeId            If scope has to be changed before this step is evaluated, id of the scope must be supplied
+    */
+  private def step[F[_], O, X](
+      stream: FreeC[Algebra[F, O, ?], Unit],
+      scopeId: Option[Token],
+      chunkSize: Int,
+      maxSteps: Long
+  ): FreeC[Algebra[F, X, ?], Option[(Segment[O, Unit], Token, FreeC[Algebra[F, O, ?], Unit])]] =
+    FreeC
+      .Eval[Algebra[F, X, ?], Option[(Segment[O, Unit], Token, FreeC[Algebra[F, O, ?], Unit])]](
+        Algebra.Step[F, O, X](stream, scopeId, chunkSize, maxSteps))
+
   def stepLeg[F[_], O](
       leg: Stream.StepLeg[F, O]): FreeC[Algebra[F, Nothing, ?], Option[Stream.StepLeg[F, O]]] =
-    FreeC.Eval[Algebra[F, Nothing, ?], Option[Stream.StepLeg[F, O]]](RunLeg(leg))
+    step[F, O, Nothing](
+      leg.next,
+      Some(leg.scopeId),
+      1024,
```

Comment:
At some point, I'd like to surface these values in to the API or otherwise make them configurable but definitely don't need to do that now.

Reply:
Although since they appear both here and in `compile`, maybe we should refactor this a bit so they are hard coded once for now?

Reply:
We can just hard code these values now to places where they are required? That way it will be the easiest to refactor once we decide how to expose these values.

Reply:
Yeah that sounds good